### PR TITLE
Adding DateTimePicker component

### PR DESCRIPTION
**Pull request by joelclimbsthings**
### All Submissions:

-   [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md)?
-   [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
-   [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change?

**To-do**
- Locales; for some reason `moment()` doesn't have the [correct output](https://momentjs.com/docs/#/displaying/) for `moment().format('LT')`, displaying `g:i pm` instead of the actual time. Better way to handle this?
- Docs and stories

### Changes proposed in this Pull Request:

Adding the DateTimePicker component.

### How to test the changes in this Pull Request:

1. Import the `DateTimePicker` component from `@woocommerce/components`
2. Throw it on a React screen.
3. Select a date via the Picker.
4. Change the date via editing the input field.

### Other information:

-   [x] Have you added an explanation of what your changes do and why you'd like us to include them?
-   [ ] Have you written new tests for your changes, as applicable?
-   [x] Have you successfully run tests with your changes locally?
-   [x] Have you created a changelog file for each project being changed, ie `pnpm changelog add --filter=<project>`?

### FOR PR REVIEWER ONLY:

-   [ ] I have reviewed that everything is sanitized/escaped appropriately for any SQL or XSS injection possibilities. I made sure Linting is not ignored or disabled.

### CODE REVIEW

File: packages/js/components/src/date-time-picker/date-time-picker.tsx [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+						if (
+							! relatedTargetParent ||
+							relatedTargetParent !== currentTargetParent
+						) {
+							onToggle();
+						}
+					} }
+					dateFormat={ dateTimeFormat }
+					label={ __( 'Choose a date', 'woocommerce' ) }
+					error={ inputError }
```
Comment (joshuatf) [inline packages/js/components/src/date-time-picker/date-time-picker.tsx]: This error display is a little odd, but is housed in `DateInput` component.  I think we should save fixing this for a follow-up if we decide to do so.

**Review by joshuatf — COMMENTED** [review]
Added some stories, readme, and reduce the prop coverage on this component a tiny bit.

This should be ready for review.

>Locales; for some reason moment() doesn't have the [correct output](https://momentjs.com/docs/#/displaying/) for moment().format('LT'), displaying g:i pm instead of the actual time. Better way to handle this?

@joelclimbsthings is on AFK and I'm not 100% sure where he was referring to, but I think any additional locale formatting may be provided by WP.

As far as this component is concerned, it should operate without the need or knowledge of WP locales.

File: packages/js/components/src/date-time-picker/date-time-picker.tsx [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+	currentDate?: string;
+	is12Hour?: boolean;
+} & React.HTMLAttributes< HTMLDivElement >;
+
+export const DateTimePicker: React.FC< DateTimeProps > = ( {
+	onChange,
+	is12Hour = true,
+	dateTimeFormat = is12Hour ? 'MM/DD/YYYY h:mm a' : 'MM/DD/YYYY H:MM',
+	disabled = false,
+	currentDate = new Date().toISOString(),
```
Comment (louwie17) [inline packages/js/components/src/date-time-picker/date-time-picker.tsx]: I think we should also add a `useEffect` for the `currentDate` prop and update the `dateTime` state when the `currentDate` updates.
So we can update it from the outside of the component as well and not just on initial render.

**Review by louwie17 — DISMISSED** [review]
Nice work @joelclimbsthings and @joshuatf, I left one suggestion for keeping an eye out on any outside changes of `currentDate` as well.

Also I ran into a small bug, which may or may not be an issue with the Wordpress date picker (most likely is).
When I select a month using the dropdown the calendar below doesn't update. If I then select a day right after, the dropdown automatically closes and only the month updates, not the newly selected day.
This is the only occurrence when it automatically closes the dropdown. See GIF:
strange calendar behaviour

**Comment by louwie17** [comment]
I am fine if the bug above is fixed at a later point, the `useEffect` would be nicer to include right away.

**Comment by joshuatf** [comment]
>When I select a month using the dropdown the calendar below doesn't update.

@louwie17 I was able to reproduce this, but after removing `node_modules` and `pnpm i` I'm not able to anymore. Not sure if that resolved something or if something else is happening.

>I am fine if the bug above is fixed at a later point, the useEffect would be nicer to include right away.

Good idea!  Added.

Also added a README for this component.

**Comment by joshuatf** [comment]
>I'm not able to anymore. Not sure if that resolved something or if something else is happening.

I was just able to reproduce this again and isolated to the GB DateTimePicker so fairly confident the issue is stemming from that component.

**Review by octaedro — APPROVED** [review]
Good work @joelclimbsthings and @joshuatf! The code looks great and is testing well here (I only saw the small bug that Lourens [mentioned before](https://github.com/woocommerce/woocommerce/pull/34293#pullrequestreview-1076192017), which I understand is not related to this PR) 🚀

**Pull request merged**